### ISSUE TITLE
Store system time at start of sampling in `CmdStanFit` objects

### ISSUE THREAD

**Issue by wpetry** [stan-dev/cmdstanr#490]
**Is your feature request related to a problem? Please describe.**
I would like to access the timestamp corresponding to the start of sampling in `CmdStanFit` objects, including those that are saved in .rds files via `$save_object()`. Information about when a model was fit has many potential uses, including for archiving model fits and for cross-referencing fits against data versions when the data are regularly updated.

`rstan` stores this information in the `date` slot of `stanfit` objects from the `rstan` package (see https://mc-stan.org/rstan/reference/stanfit-class.html).

**Describe the solution you'd like**
I would like:
1. `Sys.time()` to be called at the start of `$sampling()`
2. the resulting timestamp to be stored in the `CmdStanFit` object returned

I think the first already happens to facilitate chain timing, but `$time()` contains only information on fitting duration (not when the fitting was done). To me, `fit$metadata` would be an intuitive place to store the sampling initiation timestamp with least risk of breaking changes.

Furthermore, I think that `date` (used by `rstan`) is an ambiguous name for this timestamp—it erases the hour/minute component and isn't clear about which part of the modeling process it is stamping (compilation? sampling start? sampling end?). I suggest using a more descriptive name like "sampling_start", "sampling_timestamp", or "when_sampled." Either way, I suggest documenting that it is intended to mirror a `stanfit` `date` slot from `rstan`.

**Describe alternatives you've considered**
Some internal functions called during sampling write temporary files that include timestamps in %Y%m%d%H%M format in the file name. Perhaps these could be used to extract when samples were drawn? It's clear that this would be cumbersome, and could be prone to error in some situations.

**Additional context**
Currently, `cmdstanr` has good support for including timestamps in file names when writing components of the fitted model to disk, but these reflect when those files are written out rather than the posterior samples themselves were drawn. As far as I can tell, a user with a .rds file from `$save_object()` would not be able to determine when the sampling occurred. The difference between sampling start and file writing timestamps will likely matter most for models that take hours to days to sample. Continuous data updating will compound the problem, making it difficult for a human user to determine which version of the data was used to fit the model without extracting the data from the fitted model and comparing it to all previous data versions.

**Comment by rok-cesnovar** [comment]
I like the idea, should be simple to add.

**Comment by jgabry** [comment]
Yeah I like the idea too, thanks for the suggestion!

**Comment by jgabry** [comment]
I opened an issue about this in the cmdstan repo (https://github.com/stan-dev/cmdstan/issues/1006) because it would be ideal if we could add a timestamp to the cmdstan csv files themselves. That way cmdstanr and cmdstanpy and anyone else could access that information. For cmdstanr we could store it in metadata as @wpetry suggested. If it turns out there's opposition to adding the timestamp in cmdstan itself then we can do a cmdstanr only solution.

**Comment by rok-cesnovar** [comment]
The CmdStan PR was merged and once 2.27 will be available (in about a week or so) you will be able to do the following:

```
> fit <- cmdstanr_example("logistic")
> fit$metadata()$start_datetime
[1] "2021-05-19 06:53:52 UTC"
```

I am going to close this PR, will add a simple test as part of a PR to add other minor 2.27 features. Thanks again for the feature request!
